File: src/starfit/fitness/solver.f90
! interface module for f2py

! include and wrap only what is actually needed

subroutine fitness_(f, c, obs, err, det, cov, abu, nel, ncov, nstar, nsol, ls, icdf, flags)

  use fitting, only: &
       fitness

  use typedef, only: &
       real64, int64

  implicit none

  integer(kind=int64), intent(in) :: &
       nstar, nel, nsol, ncov

  real(kind=real64), dimension(nel), intent(in) :: &
       obs, err, det
  real(kind=real64), dimension(nel, ncov), intent(in) :: &
       cov
  real(kind=real64), dimension(nsol, nstar, nel), intent(in) :: &
       abu
  integer(kind=int64), intent(in) :: &
       ls
  integer(kind=int64), intent(in) :: &
       icdf, flags

  !f2py real(kind=real64), intent(out), dimension(nsol) :: f
  real(kind=real64), dimension(nsol), intent(out) :: &
       f
  !f2py real(kind=real64), intent(in,out), dimension(nsol, nstar) :: c
  real(kind=real64), dimension(nsol, nstar), intent(inout) :: &
       c

  call fitness(f, c, obs, err, det, cov, abu, nel, ncov, nstar, nsol, ls, icdf, flags)

end subroutine fitness_


subroutine fitness_m_(f, c, obs, err, det, cov, abu, nel, ncov, nstar, nsol, ls, icdf, flags)

  use fitting, only: &
       fitness_m

  use typedef, only: &
       real64, int64

  implicit none

  integer(kind=int64), intent(in) :: &
       nstar, nel, nsol, ncov

  real(kind=real64), dimension(nel), intent(in) :: &
       obs, err, det
  real(kind=real64), dimension(nel, ncov), intent(in) :: &
       cov
  real(kind=real64), dimension(nsol, nstar, nel), intent(in) :: &
       abu
  integer(kind=int64), intent(in) :: &
       ls
  integer(kind=int64), intent(in) :: &
       icdf, flags

  !f2py real(kind=real64), intent(out), dimension(nsol, nel, nel) :: f
  real(kind=real64), dimension(nsol, nel, nel), intent(out) :: &
       f
  !f2py real(kind=real64), intent(in,out), dimension(nsol, nstar) :: c
  real(kind=real64), dimension(nsol, nstar), intent(inout) :: &
       c

  call fitness_m(f, c, obs, err, det, cov, abu, nel, ncov, nstar, nsol, ls, icdf, flags)

end subroutine fitness_m_


subroutine get_complete_matrix_(m, obs, err, det, cov, nel, ncov, icdf)

  use star_data, only: &
       set_star_data, get_complete_matrix

  use typedef, only: &
       real64, int64

  implicit none

  integer(kind=int64), intent(in) :: &
       nel, ncov

  real(kind=real64), dimension(nel), intent(in) :: &
       obs, err, det
  real(kind=real64), dimension(nel, ncov), intent(in) :: &
       cov
  integer(kind=int64), intent(in) :: &
       icdf

  !f2py real(kind=real64), intent(out), dimension(nel, nel) :: f
  real(kind=real64), dimension(nel,nel), intent(out) :: &
       m

  call set_star_data(obs, err, det, cov, nel, ncov, icdf)
  call get_complete_matrix(m)

end subroutine get_complete_matrix_


subroutine get_complete_inverse_(m1, obs, err, det, cov, nel, ncov, icdf)

  use star_data, only: &
       set_star_data, get_complete_inverse

  use typedef, only: &
       real64, int64

  implicit none

  integer(kind=int64), intent(in) :: &
       nel, ncov

  real(kind=real64), dimension(nel), intent(in) :: &
       obs, err, det
  real(kind=real64), dimension(nel, ncov), intent(in) :: &
       cov
  integer(kind=int64), intent(in) :: &
       icdf

  !f2py real(kind=real64), intent(out), dimension(nel, nel) :: f
  real(kind=real64), dimension(nel,nel), intent(out) :: &
       m1

  call set_star_data(obs, err, det, cov, nel, ncov, icdf)
  call get_complete_inverse(m1)

end subroutine get_complete_inverse_

File: src/starfit/fitness/solver.py
import numpy as np

from ._solver import fitness_, fitness_m_, get_complete_inverse_, get_complete_matrix_


def check_offsets(offsets, flags, tag):
    if flags & FLAGS_LIMIT_SOLUTION:
        if np.any((s := np.sum(offsets, axis=-1)) > 1):
            i = np.where(s > 1)[0]
            raise AttributeError(
                f"[fitness][{tag}] invalid sum of offsets for solution(s) {i} > 1: {s[i]}"
            )
    if flags & FLAGS_LIMITED_SOLVER:
        if np.any(offsets > 1):
            ii = np.any(offsets > 1, axis=-1)
            i = np.where(ii)[0][0]
            raise AttributeError(
                f"[fitness][{tag}] invalid offset > 1 at index {i}: {offsets[i]}"
            )


ALMOST_ONE = 1.0e0 - 1.0e-14

# from f90 module
FLAGS_LIMIT_SOLUTION_BIT = 0
FLAGS_LIMITED_SOLVER_BIT = 1

# flags as addable integers
FLAGS_LIMIT_SOLUTION = 2**FLAGS_LIMIT_SOLUTION_BIT
FLAGS_LIMITED_SOLVER = 2**FLAGS_LIMITED_SOLVER_BIT

# default flags
FLAGS_DEFAULT = FLAGS_LIMIT_SOLUTION + FLAGS_LIMITED_SOLVER

# local search options
LS_CHI_ONLY = -1
LS_OFF = 0
LS_ON = 1
LS_PSOLVE = 2

LS_MODES = {
    None: LS_CHI_ONLY,
    False: LS_OFF,
    True: LS_ON,
    Ellipsis: LS_PSOLVE,
}


def fitness(
    observed,
    error,
    detection,
    covariance,
    abu,
    offsets,
    cdf=None,
    local_search=True,
    limit_solution=None,
    limit_solver=None,
    return_matrix=False,
):

    nsol = abu.shape[0]
    nstar = abu.shape[1]
    nel = abu.shape[2]
    ncov = covariance.shape[1]

    if limit_solution is None:
        limit_solution = True
    if limit_solver is None:
        limit_solver = True

    flags = 0
    if limit_solution:
        flags += FLAGS_LIMIT_SOLUTION
    if limit_solver:
        flags += FLAGS_LIMITED_SOLVER

    if cdf is None:
        cdf = True
    if cdf:
        icdf = 1
    else:
        icdf = 0

    try:
        check_offsets(offsets, flags, "INPUT")
    except:
        if flags & FLAGS_LIMIT_SOLUTION:
            s = np.sum(offsets, axis=-1)
            ii = s > ALMOST_ONE
            offsets[ii, :] = offsets[ii, :] / s[ii, np.newaxis]
        else:
            raise

    ils = LS_MODES.get(local_search, None)
    if ils is None:
        raise Exception(f"Unknown search option {local_search=}")

    if return_matrix:
        fitness_x = fitness_m_
    else:
        fitness_x = fitness_

    fitness, offset = fitness_x(
        c=offsets,
        obs=observed,
        err=error,
        det=detection,
        cov=covariance,
        abu=abu,
        nel=nel,
        ncov=ncov,
        nstar=nstar,
        nsol=nsol,
        ls=ils,
        icdf=icdf,
        flags=flags,
    )

    check_offsets(offset, flags, "RESULT")

    return fitness, offset


def get_complete_matrix(star, cdf):
    if cdf:
        icdf = 1
    else:
        icdf = 0

    m = get_complete_matrix_(
        obs=star.element_abundances.abundance,
        err=star.element_abundances.error,
        det=star.element_abundances.detection,
        cov=star.element_abundances.covariance,
        nel=star.element_abundances.shape[0],
        ncov=star.element_abundances.covariance.shape[1],
        icdf=icdf,
    )

    return m


def get_complete_inverse(star, cdf):
    if cdf:
        icdf = 1
    else:
        icdf = 0

    m = get_complete_inverse_(
        obs=star.element_abundances.abundance,
        err=star.element_abundances.error,
        det=star.element_abundances.detection,
        cov=star.element_abundances.covariance,
        nel=star.element_abundances.shape[0],
        ncov=star.element_abundances.covariance.shape[1],
        icdf=icdf,
    )

    return m
